paperqa/agents/__init__.py

```diff
@@ -0,0 +1,500 @@
+        raise KeyError(f"Invalid path: {path}")
+
+    value, should_remove = pop_nested_dict_recursive(d[keys[0]], ".".join(keys[1:]))
+
+    if should_remove:
+        d.pop(keys[0])
+
+    return value, len(d) == 0
+
+
+def get_current_settings(
+    settings_path: str | os.PathLike | None = None,
+) -> dict[str, Any]:
+
+    if settings_path is None:
+        settings_path = pqa_directory("settings") / "settings.yaml"
+
+    if os.path.exists(settings_path):
+        with open(settings_path) as f:
+            return yaml.safe_load(f)
+
+    return {}
+
+
+def merge_dicts(dict_a: dict, dict_b: dict) -> dict:
+    """
+    Merge two dictionaries where if dict_a has a key with a subdictionary.
+
+    dict_b only overwrites the keys in dict_a's subdictionary if they are
+    also specified in dict_b, but otherwise keeps all the subkeys.
+    """
+    result = dict_a.copy()  # Start with a shallow copy of dict_a
+
+    for key, value in dict_b.items():
+        if isinstance(value, dict) and key in result and isinstance(result[key], dict):
+            # If both dict_a and dict_b have a dict for this key, recurse
+            result[key] = merge_dicts(result[key], value)
+        else:
+            # Otherwise, just update the value
+            result[key] = value
+
+    return result
+
+
+def get_merged_settings(
+    settings: dict[str, Any], settings_path: Path | None = None
+) -> dict[str, Any]:
+    """Merges a new settings with the current settings saved to file."""
+    if settings_path is None:
+        settings_path = pqa_directory("settings") / "settings.yaml"
+
+    current_settings = get_current_settings(settings_path)
+
+    # deal with the nested key case
+    return merge_dicts(current_settings, settings)
+
+
+@app.command("set")
+def set_setting(
+    variable: Annotated[
+        str,
+        typer.Argument(
+            help=(
+                "PaperQA variable to set, see agents.models.QueryRequest object for all settings, "
+                "nested options can be set using periods, ex. agent_tools.paper_directory"
+            )
+        ),
+    ],
+    value: Annotated[
+        str,
+        typer.Argument(
+            help=(
+                "Value to set to the variable, will be cast to the correct type automatically."
+            )
+        ),
+    ],
+) -> bool:
+    """Set a persistent PaperQA setting."""
+    settings_path = pqa_directory("settings") / "settings.yaml"
+
+    current_settings = get_merged_settings(
+        parse_dot_to_dict(variable, value), settings_path=settings_path
+    )
+
+    try:
+        _ = QueryRequest(**current_settings)
+    except MismatchedModelsError:
+        pass
+    except ValueError as e:
+        raise ValueError(
+            f"{variable} (with value {value}) is not a valid setting."
+        ) from e
+
+    logger.info(f"{variable} set to {str(value)[:100]}!")
+
+    with open(settings_path, "w") as f:
+        yaml.dump(current_settings, f)
+
+    return True
```

Comment:
Let's not return anything, in validators it's best to:
- Fail and blow up with an `Exception`
- Silently succeed on success